app/server/lib/BootProbes.ts

```diff
@@ -0,0 +1,192 @@
+import { ApiError } from 'app/common/ApiError';
+import { BootProbeIds, BootProbeResult } from 'app/common/BootProbe';
+import { removeTrailingSlash } from 'app/common/gutil';
+import { expressWrap, jsonErrorHandler } from 'app/server/lib/expressWrap';
+import { GristServer } from 'app/server/lib/GristServer';
+import * as express from 'express';
+import fetch from 'node-fetch';
+
+/**
+ * Self-diagnostics useful when installing Grist.
+ */
+export class BootProbes {
+  // List of probes.
+  public _probes = new Array<Probe>();
+
+  // Probes indexed by id.
+  public _probeById = new Map<string, Probe>();
+
+  public constructor(private _app: express.Application,
+                     private _server: GristServer,
+                     private _base: string) {
+    this._addProbes();
+  }
+
+  public addEndpoints() {
+    // Return a list of available probes.
+    this._app.use(`${this._base}/probe$`, expressWrap(async (_, res) => {
+      res.json({
+        'probes': this._probes.map(probe => {
+          return { id: probe.id, name: probe.name };
+        }),
+      });
+    }));
+
+    // Return result of running an individual probe.
+    this._app.use(`${this._base}/probe/:probeId`, expressWrap(async (req, res) => {
+      const probe = this._probeById.get(req.params.probeId);
+      if (!probe) {
+        throw new ApiError('unknown probe', 400);
+      }
+      const result = await probe.apply(this._server, req);
+      res.json(result);
+    }));
+
+    // Fall-back for errors.
+    this._app.use(`${this._base}/probe`, jsonErrorHandler);
+  }
+
+  private _addProbes() {
+    this._probes.push(_homeUrlReachableProbe);
+    this._probes.push(_statusCheckProbe);
+    this._probes.push(_userProbe);
+    this._probes.push(_bootProbe);
+    this._probes.push(_hostHeaderProbe);
+    this._probeById = new Map(this._probes.map(p => [p.id, p]));
+  }
+}
+
+/**
+ * An individual probe has an id, a name, an optional description,
+ * and a method that returns a probe result.
+ */
+export interface Probe {
+  id: BootProbeIds;
+  name: string;
+  description?: string;
+  apply: (server: GristServer, req: express.Request) => Promise<BootProbeResult>;
+}
+
+const _homeUrlReachableProbe: Probe = {
+  id: 'reachable',
+  name: 'Grist is reachable',
+  apply: async (server, req) => {
+    const url = server.getHomeUrl(req);
+    try {
+      const resp = await fetch(url);
+      if (resp.status !== 200) {
+        return {
+          success: false,
+          severity: 'fault',
+          details: {
+            error: await resp.text(),
+            status: resp.status,
+          }
+        };
```

Comment:
This and the catch block are very similar. Any chance we can re-use it by storing the relevant bits in an `ApiError` and throwing it ourselves?